src/intervals/construction.jl

```diff
@@ -66,20 +68,26 @@ constructors:
     are included in the interval even if they can not be represented as
     floating point numbers.
 """
-struct Interval{T} <: Real
+struct Interval{T<:NumTypes} <: Real
     lo::T
     hi::T
 
-    Interval{T}(a::T, b::T) where {T} = new{T}(_normalisezero(a), _normalisezero(b))
+    Interval{T}(a::T, b::T) where {S<:Integer,T<:Rational{S}} = new{T}(_normalisezero(a), _normalisezero(b))
+    Interval{T}(a::T, b::T) where {T<:AbstractFloat} = new{T}(_normalisezero(a), _normalisezero(b))
+end
+
+function Interval{T}(a, b) where {S<:Integer,T<:Rational{S}}
+    R = float(S)
+    return Interval{T}(rationalize(S, prevfloat(R(a, RoundDown))), rationalize(S, nextfloat(R(b, RoundUp))))
 end
 
-Interval{T}(a, b) where {T} = Interval{T}(T(a, RoundDown), T(b, RoundUp))
+Interval{T}(a, b) where {T<:AbstractFloat} = Interval{T}(T(a, RoundDown), T(b, RoundUp))
```

Comment:
Just a note that rounding, here, is not working:
```julia
julia> setformat(:full); 

julia> Interval{Float64}(0.1, 0.1)
Interval(0.1, 0.1)

julia> Interval{BigFloat}(big(0.1), big(0.1))
Interval(0.1000000000000000055511151231257827021181583404541015625, 0.1000000000000000055511151231257827021181583404541015625)
```

Reply:
Yes `interval` has become the only correctly rounded interval constructor. In my mind, any call to `Interval` is unsafe.

Reply:
i think the same applies to `interval`:
```julia
julia> setformat(:full); 

julia> Interval{Float64}(0.1, 0.1)
Interval{Float64}(0.1, 0.1)

julia> interval(0.1, 0.1)
Interval{Float64}(0.1, 0.1)
```

Reply:
Anyway, this should be left for another PR....

Reply:
Ah indeed. But is that something we want to change? I mean `0.1` is a float in its own right and calling `interval` creates a valid interval that encapsulates that float after all.

Reply:
I agree that `0.1` is a `Float64`, indeed, but it includes already rounding , and it is actually rounded up (the simplest way to see this is `big(0.1)`). Then, the interval created does not contain the real `0.1 == 1//10`; the interval is indeed tight, but it's definitely not valid. 
This is not the motivation for this PR, yet it is an important concern.
cc @Kolaru

Reply:
Absolutely. Honestly, I do not know how to disambiguate the situation (besides the use of strings and macros that we already have). When `interval` is being called, Julia only has access to the (already parsed) float without knowing that the user actually asked for some nearby value (e.g. the float `0.1` vs the actual value 1/10); in this regard, I consider the current behaviour valid.
Unless perhaps it is possible to monitor Julia's parsing system?

Reply:
I see. In fact why was it chosen that `atomic` does not return the tightest interval?
I may be overlooking something here, but replacing the lines https://github.com/JuliaIntervals/IntervalArithmetic.jl/blob/4d6bd155a2d92cc442113c3feba36a7b7b17844b/src/intervals/construction.jl#L209-L214 with
```Julia
if x == lo == hi
    lo = prevfloat(lo)
end
```
would give us the tightest valid interval, no?

Reply:
Not in general... It works for `0.1`, because the parsed `0.1` is rounded upwards. If you use `0.3`, which is rounded downwards, it wouldn't work.... In other words, the Float64 "0.1" represents a real slightly larger than 1//10, and the Float64 `0.3` is a real slightly smaller than 3//10.
The problem is how to know the actual rounding mode used when parsing, to compensate it...

Reply:
Ah indeed, thanks for the clarification 🙂

Reply:
Even with a macro, we can not intercept the input before parsing to Float64. So this issue can only be solved in few ways
1. Just accept that the Float64 we get is the one intended (current state)
2. Use strings and parse them ourselves
3. Widen all floats as `(prefloat(x)..nexfloat(x))` (as Luis explained we can not do better)
4. Try to guess what the user intended (that use to be the case, converting floats to the closest reasonnable rational number)

At some point it was decided to remove the 4th possibility, and let the user decide what they want. The documentation still needs some big warning about that somewhere though.